draft-ietf-quic-transport.md

```diff
@@ -5060,8 +5059,8 @@ the offset of the next byte that would be sent.
 The first byte in the stream has an offset of 0.  The largest offset delivered
 on a stream - the sum of the offset and data length - cannot exceed 2^62-1, as
 it is not possible to provide flow control credit for that data.  Receipt of a
-frame that exceeds this limit will be treated as a connection error of type
-FLOW_CONTROL_ERROR.
+frame that exceeds this limit MUST be treated as a connection error of type
+FRAME_ENCODING_ERROR or FLOW_CONTROL_ERROR.
```

Comment:
I'd be inclined towards only recommending FLOW_CONTROL_ERROR here.

Reply:
I think that permitting either is in line with principles.  I mean, sure, this can be flow control, but if someone has code that validates frames and that code all directs into a FRAME_ENCODING_ERROR code path, that's fine.  Checking that `x+y>=2^62` might be redundant, but I don't want to say that that code can't run that check.

Reply:
I don't think options are helpful here, so I'd prefer the more specific error.

Reply:
If there is a flow control limit < 2^62 then presumably you would check against that and not a theoretical max limits. Thus having two different errors would needlessly add an extra check.

Reply:
Note that there is a similar issue with MAX_STREAMS and STREAMS_BLOCKED. They communicate the cumulative count of streams of particular type using varint. It is possible to express numbers up to 2<sup>62</sup> however it is always an error to indicate a value above 2<sup>60</sup>.

[Section 4.5](https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#rfc.section.4.5) currently states that STREAM_LIMIT_ERROR is to be used for communicating such error (for MAX_STREAMS, there is no mention of STREAMS_BLOCKED).

Reply:
That was done before this proposed principle.  Do you think that we should be aiming to have more specific error codes where possible here?

Reply:
My point is that if we are going to permit the use of FRAME_ENCODING_ERROR when the STREAM frame exceeds the 2<sup>62</sup> boundary, then we should also permit the use of FRAME_ENCODING_ERROR for MAX_STREAMS frame carrying a count exceeding 2<sup>60</sup>.

My weak preference is to not do either of the two, as we'd be losing a clear signal. IMO, this discussion is about the error code to be used when integrity check for certain frames fail. We can argue that integrity check is not part of frame decoding. Then, it would be natural to use the more specific error codes.